File: python/sparklabs/batch/sales/sql/orders_revenue_case_when.py
from sparklabs.batch.sales.sql import utils


def run() -> None:
    utils.sql(
        """
        WITH revenue_base AS (
          SELECT oi.order_id,
                 p.price * oi.quantity AS gross_revenue,
                 CASE
                   WHEN TO_DATE(o.order_date) BETWEEN TO_DATE(pr.start_date) AND TO_DATE(pr.end_date)
                     THEN COALESCE(pr.discount_pct, 0)
                   ELSE 0
                 END AS discount_pct
          FROM order_items oi
          JOIN orders o ON oi.order_id = o.order_id
          JOIN products p ON oi.product_id = p.product_id
          LEFT JOIN promotions pr ON p.category = pr.category
        )
        SELECT order_id,
               ROUND(SUM(gross_revenue), 2) AS gross_revenue,
               ROUND(SUM(gross_revenue * discount_pct / 100), 2) AS discount_value,
               ROUND(SUM(gross_revenue - (gross_revenue * discount_pct / 100)), 2) AS net_revenue
        FROM revenue_base
        GROUP BY order_id
        ORDER BY net_revenue DESC
        """
    ).show(truncate=False)


if __name__ == "__main__":
    run()

File: r/sparklabs/batch/sales/dataframe/orders_revenue_case_when.R
source("r/sparklabs/batch/sales/dataframe/utils.R")

base_df <- join(order_items, orders, order_items$order_id == orders$order_id)
base_df <- join(base_df, products, base_df$product_id == products$product_id)
base_df <- join(base_df, promotions, base_df$category == promotions$category, "left")
base_df <- withColumn(base_df, "gross_revenue", base_df$price * base_df$quantity)
base_df <- withColumn(
  base_df,
  "discount_pct_active",
  ifelse(
    to_date(base_df$order_date) >= to_date(base_df$start_date) & to_date(base_df$order_date) <= to_date(base_df$end_date),
    coalesce(base_df$discount_pct, lit(0)),
    lit(0)
  )
)
base_df <- withColumn(base_df, "discount_value", base_df$gross_revenue * base_df$discount_pct_active / 100)
base_df <- withColumn(base_df, "net_revenue", base_df$gross_revenue - base_df$discount_value)

result <- orderBy(
  agg(
    groupBy(base_df, base_df$order_id),
    gross_revenue = round(sum(base_df$gross_revenue), 2),
    discount_value = round(sum(base_df$discount_value), 2),
    net_revenue = round(sum(base_df$net_revenue), 2)
  ),
  desc("net_revenue")
)
showDF(result, truncate = FALSE)
